Introduce vsock support for attestation

luigix25 opened the pull request:
This PR introduces the vsock transport for the attestation.

Vsock is more flexible than the serial port as it has the concept of port, and we can reuse vsock for anything else.

With this PR SVSM tries to perform attestation first with vsock, if it fails it falls back using the serial port.
No change intended to the attestation's logic, I am just adding a new transport type.

Default vsock port for attestation is `1995`. 

To compile it you need to enable the feature `attest` and  `vsock`.  To run it follow the instructions available in the Documentation.

luigix25 committed:
aproxy: use read/write traits

The attestation functions take concrete UnixStream parameters, which
prevents reusing them with other transports such as vsock.

Change parameter types from UnixStream to impl Read + Write to
decouple the attestation logic from the concrete stream type.

No functional change intended.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
aproxy: use read/write traits

The attestation functions take concrete UnixStream parameters, which
prevents reusing them with other transports such as vsock.

Change parameter types from UnixStream to impl Read + Write to
decouple the attestation logic from the concrete stream type.

No functional change intended.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
aproxy: enable vsock for attestation

The attestation proxy currently only supports UNIX domain sockets.

Add vsock support for attestation as it's now available in SVSM.

This can be used via the --vsock option to listen on a vsock port
(default: 1995) instead of a UNIX domain socket.

The two transports are mutually exclusive.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
aproxy: enable vsock for attestation

The attestation proxy currently only supports UNIX domain sockets.

Add vsock support for attestation as it's now available in SVSM.

This can be used via the --vsock option to listen on a vsock port
(default: 1995) instead of a UNIX domain socket.

The two transports are mutually exclusive.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/attest: abstract transport implementation

The attestation driver directly uses a `SerialPort`, making it
impossible to use a different communication channel without
modifying the driver code.

Introduce a `Transport` enum that wraps the underlying channel and
implements `Read` and `Write`. This allows adding new transports such
as virtio-vsock by simply adding enum variants, without changing the
attestation protocol code.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/attest: abstract transport implementation

The attestation driver directly uses a `SerialPort`, making it
impossible to use a different communication channel without
modifying the driver code.

Introduce a `Transport` enum that wraps the underlying channel and
implements `Read` and `Write`. This allows adding new transports such
as virtio-vsock by simply adding enum variants, without changing the
attestation protocol code.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/attest: add vsock transport with serial fallback

Add vsock as a `Transport` variant to support both vsock and serial
port communication. When the vsock feature is enabled, the attestation
driver will attempt to connect via vsock first on port 1995, falling back
to the serial port (COM3) on failure.

Since attestation may now use vsock, virtio-mmio initialization
needs to be performed earlier than the attestation protocol.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/attest: add vsock transport with serial fallback

Add vsock as a `Transport` variant to support both vsock and serial
port communication. When the vsock feature is enabled, the attestation
driver will attempt to connect via vsock first on port 1995, falling back
to the serial port (COM3) on failure.

Since attestation may now use vsock, virtio-mmio initialization
needs to be performed earlier than the attestation protocol.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
Documentation/ATTESTATION: document vsock transport option

Document the vsock transport mechanism as an alternative to the
serial port for attestation. Include configuration details and usage
examples.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
Documentation/ATTESTATION: document vsock transport option

Document the vsock transport mechanism as an alternative to the
serial port for attestation. Include configuration details and usage
examples.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 commented:
~Depends on #766~

luigix25 commented on aproxy/src/main.rs:
```diff
@@ -26,32 +27,46 @@ struct Args {
-    #[clap(long, short, default_value_t = false)]
+    #[clap(long, short, requires = "unix", default_value_t = false)]
     force: bool,
 }
```
I didn't know about clap groups, thanks!

luigix25 replied:
I guess there is some kind of bug in the clap help message.

With your suggested changes
```shell
bin/aproxy --vsock-port 1234 
error: the following required arguments were not provided:
  --url <URL>
  --protocol <BACKEND>
  --unix <UNIX>
```

Without 

```shell
bin/aproxy_old --vsock-port 1234         
error: the following required arguments were not provided:
  --url <URL>
  --protocol <BACKEND>
```

First one doesn't make too much sense. as `vsock-port` and `unix` cannot coexist.

Clap doesn't allow to use them together, but the help message is wrong

00xc replied:
Right, it seems groups have even more bugs :( feel free to discard my suggestion, we can make it work later if we find a better way

luigix25 committed:
io: add `read_exact` and `write_all` to Read and Write trait

Stream-oriented transports such as VsockStream may return fewer bytes
than requested from a single read() or write() call. This means that
callers need to handle partial transfers themselves.

Add `read_exact()` and `write_all()` implementations to the Read and
Write traits that loop until the full buffer is consumed, returning
an error on unexpected EOF.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
io: add `read_exact` and `write_all` to Read and Write trait

Stream-oriented transports such as VsockStream may return fewer bytes
than requested from a single read() or write() call. This means that
callers need to handle partial transfers themselves.

Add `read_exact()` and `write_all()` implementations to the Read and
Write traits that loop until the full buffer is consumed, returning
an error on unexpected EOF.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 commented:
- Rebased to latest main to pickup vsock changes.
- Added `read_exact` and `write_all` in the Read/Write trait

stefano-garzarella commented on kernel/src/error.rs:
```diff
@@ -148,6 +148,8 @@ pub enum SvsmError {
     /// Errors related to vsock.
     #[cfg(feature = "vsock")]
     Vsock(VsockError),
+    /// Unexpected end of input.
```
EoI or EoF ?

stefano-garzarella replied:
Also not sure if we should introduce IO errors.

luigix25 replied:
`read_exact` should return an error and I don't think we already have an "appropriate" error here, this is why I added a new one.

stefano-garzarella replied:
yes, that part is clear, my 2 questions are:
- why calling it Eof and have the description talking about `end of input` ?
- should we introduce `IO` or `InputOutput` errors like for example we did for `Vsock`, `Apic`, `Attestation`, etc. and move `Eof` under that type ?

luigix25 replied:
>why calling it Eof and have the description talking about end of input ?

at the beginning it was end of input, I then changed to EoF but forgot to update the doc.

>should we introduce IO or InputOutput errors like for example we did for Vsock, Apic, Attestation, etc. and move Eof under that type ?

Right

stefano-garzarella commented on tools/aproxy/src/main.rs:
```diff
@@ -24,11 +25,19 @@ struct Args {
+    )]
+    unix: Option<String>,
+
+    /// vsock port used by SVSM for attestation
```
Can we be a bit more precise here?

IIUC this is the port where aproxy is listening and SVSM will connect, right?

So maybe `vsock listening port where SVSM will connect` something like that. We don't need `for attestation` is implicit since this is the attestation proxy...

stefano-garzarella commented on tools/aproxy/src/main.rs:
```diff
@@ -24,11 +25,19 @@ struct Args {
+
+    /// vsock port used by SVSM for attestation
+    #[clap(long, conflicts_with_all = ["unix", "force"], required_unless_present("unix"))]
+    vsock_port: Option<u32>,
```
Should we add a default also here? I mean the kernel uses 1995 as default, why not having the same here?

maybe the param can be `--vsock PORT` where PORT is optional and default to 1995. Is it doable with clap?

luigix25 replied:
will do.

stefano-garzarella commented on kernel/src/io.rs:
```diff
@@ -70,11 +70,35 @@ pub trait Read {
 
     fn read(&mut self, buf: &mut [u8]) -> Result<usize, Self::Err>;
+
+    fn read_exact(&mut self, buf: &mut [u8]) -> Result<(), SvsmError> {
```
Can we use `Self::Err` like for other API ?

stefano-garzarella commented on kernel/src/io.rs:
```diff
@@ -70,11 +70,35 @@ pub trait Read {
 
     fn write(&mut self, buf: &[u8]) -> Result<usize, Self::Err>;
+
+    fn write_all(&mut self, buf: &[u8]) -> Result<(), SvsmError> {
```
Ditto about the error

stefano-garzarella commented on tools/aproxy/src/main.rs:
```diff
@@ -43,22 +52,24 @@ enum ArgsBackend {
+        for stream in listener.incoming() {
+            let mut stream = stream.context("Failed to connect to the vsock socket")?;
+            let mut http_client = backend::HttpClient::new(args.url.clone(), args.backend.into())?;
+            attest::attest(&mut stream, &mut http_client)?;
```
Can we introduce an helper for this to avoid code duplication?

luigix25 replied:
good question, but I don't think it's easily doable: 

The problem is that `VsockListener` and `UnixListener` return two different types and there is no "Stream" Trait. The only option is to create something similar to what I did with `Transport`. wdyt?

Without any changes I think we could create a function to group these two calls
```            
let mut http_client = backend::HttpClient::new(args.url.clone(), args.backend.into())?;
attest::attest(&mut stream, &mut http_client)?;
```

I don't know it it's worth it or not

stefano-garzarella replied:
> I don't know it it's worth it or not

Maybe no.

Or we can try a macro. Anyway, feel free to ignore if it will be too complex, we can do it later.

luigix25 replied:
Ok, I'll keep this open for now so we can get more feedbacks.

luigix25 replied:
I ended up using a macro. closing this

stefano-garzarella commented on kernel/src/attest.rs:
```diff
@@ -34,21 +37,59 @@ use sha2::{Digest, Sha512};
         match self {
+            #[cfg(feature = "vsock")]
+            Transport::Vsock(vsock) => {
+                vsock.write_all(buf)?;
```
mmm, it seems strange to me call write_all() inside a write().
IMO we should implement/use write_all() also for the serial and call this method write_all()

stefano-garzarella commented on kernel/src/attest.rs:
```diff
@@ -34,21 +37,59 @@ use sha2::{Digest, Sha512};
         match self {
+            #[cfg(feature = "vsock")]
+            Transport::Vsock(vsock) => {
+                vsock.read_exact(buf)?;
```
Ditto about read/read_exact()

stefano-garzarella commented on kernel/src/attest.rs:
```diff
@@ -34,21 +37,59 @@ use sha2::{Digest, Sha512};
+            Ok(value) => Transport::Vsock(value),
+            Err(e) => {
+                log::warn!(
+                    "Vsock error: {e:?} occurred when attempting to start attestation on port {ATTEST_DEFAULT_VSOCK_PORT}. Attempting to start again using the serial port.",
```
the message is unclear, we should make more clear that the connection is failing. In this way it seems that we are listening.

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -543,6 +543,9 @@ fn svsm_init(launch_info: &KernelLaunchInfo) {
         panic!("Failed to prepare guest FW: {e:#?}");
     }
 
+    #[cfg(feature = "virtio-drivers")]
+    initialize_virtio_mmio(&boot_params).expect("Failed to initialize virtio-mmio drivers");
+
```
Why this change? This should at least mentioned in the commit description

luigix25 replied:
attestation (may) require vsock, so we need to initialize it before it. Will add it to the description

stefano-garzarella commented on kernel/src/types.rs:
```diff
@@ -99,3 +99,6 @@ pub const TPR_NORMAL: usize = 0;
 pub const TPR_IPI: usize = 14;
+
+pub const ATTEST_DEFAULT_VSOCK_PORT: u32 = 1995;
+pub const ATTEST_DEFAULT_SERIAL_IO_ADDR: u16 = 0x3e8; // COM3
```
not sure if this is the best place, but I need to think more about this. Any other option?

luigix25 replied:
very good question. Maybe within `attest.rs` itself? I don't expect these consts be used anywhere else

stefano-garzarella replied:
yep, maybe that place is better. those are more default values than types, but I hope we can remove those with the device tree, so let's move them to attest.rs for now.

stefano-garzarella reviewed (commented):
Did a very quick review and left some comments. Thanks

stefano-garzarella commented on Documentation/docs/developer/ATTESTATION.md:
```diff
@@ -283,28 +315,33 @@ SEV-SNP machine with an SVSM-enabled kernel.
 
-     * `--url http://0.0.0.0:8080`: The attestation server is running at
-       `http://0.0.0.0:8080`.
-     * `--protocol kbs-test`: The attestation server communicates via the KBS
```
is it okay to change this from `kbs-test` to `kbs` ?

luigix25 replied:
I don't think `kbs-test` is supported anymore in aproxy

stefano-garzarella replied:
So, maybe better to fix it in a separate commit or at least mention it in the commit.
I thought was a rebase issue.

luigix25 committed:
kernel/attest: switch to write_all/read_exact

The attestation read/write helpers issue a single read() or write()
call, which may return fewer bytes than requested on stream-oriented
transports such as vsock. Use read_exact() and write_all() to ensure
the full buffer is transferred.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/attest: switch to write_all/read_exact

The attestation read/write helpers issue a single read() or write()
call, which may return fewer bytes than requested on stream-oriented
transports such as vsock. Use read_exact() and write_all() to ensure
the full buffer is transferred.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
Documentation/ATTESTATION: replace removed 'kbs-test' protocol with 'kbs'

The 'kbs-test' protocol option has been removed from the attestation
proxy. Update the documentation to use 'kbs', which is now the only
supported KBS protocol option.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 commented:
addressed @stefano-garzarella's comments
aproxy:
- Parameter renamed `--vsock`
- Some changes to clap. Hope it looks cleaner now. Unfortunately we sill need `conflicts_with = "vsock"` [@00xc]
- using a macro to simplify the code

svsm:
- Now using read_exact/write_all everywhere.
- Introduced IoError
- read_exact/write_all now use Self::Err

Improved commit messages.

stefano-garzarella commented on kernel/src/attest.rs:
```diff
@@ -184,7 +213,7 @@ impl AttestationDriver<'_> {
-            self.sp
+            self.transport
                 .read(&mut bytes)
                 .or(Err(AttestationError::ProxyRead))?;
```
all of this `.or()` are pre-existings, so not an issue for this PR, but I'd like to avoid it and maybe extend `ProxyRead`/`ProxyWrite` to wrap the source error

luigix25 replied:
makes sense, yes!

stefano-garzarella commented on kernel/src/attest.rs:
```diff
@@ -52,11 +62,41 @@ impl Write for Transport<'_> {
+                log::warn!(
+                    "Failed to connect to attestation proxy on vsock port \
+                     {ATTEST_DEFAULT_VSOCK_PORT}: {e:?}. \
+                     Attempting to start again using the serial port.",
```
nit: what about:
`Falling back to serial port transport.`

stefano-garzarella commented on tools/aproxy/src/main.rs:
```diff
@@ -40,26 +46,30 @@ enum ArgsBackend {
+            attest::attest(&mut stream, &mut http_client)?;
+        }
+    };
+}
```
What about avoiding the macro in this way:
```diff
@@ -46,14 +50,17 @@ enum ArgsBackend {
     Kbs,
 }
 
-macro_rules! accept_loop {
-    ($listener:expr, $url:expr, $backend:expr) => {
-        for stream in $listener.incoming() {
-            let mut stream = stream.context("Failed to accept connection")?;
-            let mut http_client = backend::HttpClient::new($url.clone(), $backend.into())?;
-            attest::attest(&mut stream, &mut http_client)?;
-        }
-    };
+fn accept_loop<S: Read + Write>(
+    incoming: impl Iterator<Item = io::Result<S>>,
+    url: &str,
+    backend: ArgsBackend,
+) -> anyhow::Result<()> {
+    for stream in incoming {
+        let mut stream = stream.context("Failed to accept connection")?;
+        let mut http_client = backend::HttpClient::new(url.to_string(), backend.into())?;
+        attest::attest(&mut stream, &mut http_client)?;
+    }
+    Ok(())
 }
 
 fn main() -> anyhow::Result<()> {
@@ -62,14 +69,14 @@ fn main() -> anyhow::Result<()> {
     if let Some(port) = args.vsock {
         let listener = VsockListener::bind(&VsockAddr::new(VMADDR_CID_ANY, port))
             .context("bind and listen failed")?;
-        accept_loop!(listener, args.url, args.backend);
+        accept_loop(listener.incoming(), &args.url, args.backend)?;
     } else if let Some(unix) = args.unix {
         if args.force {
             let _ = fs::remove_file(&unix);
         }
 
         let listener = UnixListener::bind(unix).context("unable to bind to UNIX socket")?;
-        accept_loop!(listener, args.url, args.backend);
+        accept_loop(listener.incoming(), &args.url, args.backend)?;
     }
 
     Ok(())
```

luigix25 replied:
nice, thanks!

stefano-garzarella commented on kernel/src/attest.rs:
```diff
@@ -33,7 +35,13 @@ use serde::Serialize;
 use zerocopy::{FromBytes, IntoBytes};
 
+#[cfg(feature = "vsock")]
+const ATTEST_DEFAULT_VSOCK_PORT: u32 = 1995;
```
What about defining this in `libaproxy` so both this code and `aproxy` can use the same constant?

luigix25 replied:
this makes sense. Unfortunately clap requires a string for `default_missing_value`, which is the only place where this value is used in aproxy. 

I see two options here:
- defining a "companion" string like `ATTEST_DEFAULT_VSOCK_PORT_STR`
- Use `const_format` to convert the u32 to a string at compile time

I have a slight preference over the second one.

stefano-garzarella replied:
I'm not sure that adding a new dependency just for this is a good idea.

stefano-garzarella replied:
BTW not a strong opinion on my side

luigix25 replied:
yeah, it's not super nice, but the alternatives are not nice either. We can change this later on if this becomes a packaging problem.

stefano-garzarella commented on Documentation/docs/developer/ATTESTATION.md:
Since we are here, can we replace also `--backend` with `--protocol` ? (if you want to do in another PR, it's also fine)

luigix25 replied:
yep, I'll open another PR fixing all these small problems!

luigix25 replied:
See #1046

stefano-garzarella commented on Documentation/docs/developer/ATTESTATION.md:
```diff
@@ -283,28 +315,33 @@ SEV-SNP machine with an SVSM-enabled kernel.
-     * `--force`: Remove the `/tmp/svsm-proxy.sock` file (if it already exists)
-       before creating the socket.
+    * `--unix /tmp/svsm-proxy.sock`: Listen for messages from SVSM on a socket
+      created in file `/tmp/svsm-proxy-sock`.
```
Pre-existing, but since you are touching this code, this should be `/tmp/svsm-proxy.sock`

luigix25 replied:
opened #1046

stefano-garzarella reviewed (commented):
Overall LGTM, just some minor comments.

luigix25 committed:
aproxy: factor out accept loop to a separate function

Factor out the accept loop to a separate function, that makes no
assumption on the stream type.

Also, replace the panic on accept failure with proper error propagation.

This is preparation work to support different type of streams,
like `VsockStream`.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 commented:
Addressed all @stefano-garzarella's comments:
- Moved ATTEST_DEFAULT_VSOCK_PORT to libaproxy so it consumed by aproxy and kernel
- minor changes to aproxy: using a function instead of a macro for stream handling
- rebased to latest main

stefano-garzarella commented on tools/aproxy/src/main.rs:
```diff
@@ -10,14 +10,18 @@ mod backend;
 use anyhow::Context;
 use clap::{Parser, ValueEnum};
+use const_format::formatcp;
+use libaproxy::ATTEST_DEFAULT_VSOCK_PORT;
```
This is added later by another commit...

luigix25 replied:
oops, moving it to `aproxy: enable vsock for attestation`

stefano-garzarella commented on kernel/src/attest.rs:
```diff
@@ -70,11 +108,6 @@ impl TryFrom<Tee> for AttestationDriver<'_> {
     type Error = SvsmError;
 
     fn try_from(tee: Tee) -> Result<Self, Self::Error> {
-        // TODO: Make the IO port configurable/discoverable for other transport mechanisms such as
```
Please don't remove this TODO, maybe move this near the new `ATTEST_DEFAULT_SERIAL_IO_ADDR` declaration.

stefano-garzarella commented on Documentation/docs/developer/ATTESTATION.md:
````diff
@@ -283,28 +315,33 @@ SEV-SNP machine with an SVSM-enabled kernel.
+      ```shell
+      ./scripts/launch_guest.sh --qemu $QEMU \
+                                --image $QCOW2 \
+                                --aproxy /tmp/svsm-proxy.sock
````
maybe we should rename `--aproxy` to `aproxy-serial` or extend it to support both serial or vsock, anyway we can fix it later.

luigix25 replied:
>extend it to support both serial or vsock, anyway we can fix it later

what do you mean? for vsock we just need to attach the device (`--vsock cid`)

stefano-garzarella replied:
I mean that `--aproxy vsock:3` can attach a vsock device with CID = 3, similar to `--aproxy serial:/path/to/unix` that can attach a serial port with the unix backend. IMO is more uniform.

luigix25 replied:
uh ok, will do in a followup PR.

stefano-garzarella reviewed (changes requested):
We should fix the bisection breakage in commit 4, the rest LGTM. I left some minor comments, the one related to the script can be fixed later.

stefano-garzarella commented on kernel/src/attest.rs:
```diff
@@ -33,11 +35,73 @@ use serde::Serialize;
 use zerocopy::{FromBytes, IntoBytes};
 
+// TODO: Make the IO port configurable/discoverable for other transport mechanisms such as
+// virtio-vsock
```
I'd remove the vsock mention here and also I'd mention that maybe we can remove the support for serial port at all

stefano-garzarella commented:
@tylerfanelli can you take a look?

tylerfanelli reviewed (approved):
LGTM, thanks!
